### PR TITLE
Minimize amount of fake `DefId`s used in rustdoc

### PR DESCRIPTION
**Pull request by Stupremee**
Follow up from #84707, which minimizes the amount of fake defids to the smallest amount possible. Every `FakeDefId` that is now used in the rustdoc library must be preserved and can not be replaced with a normal `DefId`.

### CODE REVIEW

File: src/librustdoc/json/conversions.rs [hunk 1]
```diff
@@ -546,13 +545,13 @@ impl FromWithTcx<clean::Import> for Import {
             Simple(s) => Import {
                 source: import.source.path.whole_name(),
                 name: s.to_string(),
-                id: import.source.did.map(FakeDefId::from).map(from_def_id),
+                id: import.source.did.map(Into::into).map(from_def_id),
```
Comment (jyn514) [inline src/librustdoc/json/conversions.rs:548]: Why this change?

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 2]
```diff
@@ -213,6 +213,6 @@ impl BadImplStripper {
     }
 
     fn keep_impl_with_def_id(&self, did: &FakeDefId) -> bool {
-        self.items.contains(did)
+        self.items.contains(&did)
```
Comment (jyn514) [inline src/librustdoc/passes/collect_trait_impls.rs:216]: Why this change?

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 3]
```diff
@@ -213,6 +213,6 @@ impl BadImplStripper {
     }
 
     fn keep_impl_with_def_id(&self, did: &FakeDefId) -> bool {
```
Comment (jyn514) [inline src/librustdoc/passes/collect_trait_impls.rs:215]: Passing a reference seems weird here  - why not pass by value?

**Review by jyn514 — COMMENTED** [review]
This is great :heart:

File: src/librustdoc/json/conversions.rs (hunk 1 above)
Comment (Stupremee) [reply src/librustdoc/json/conversions.rs:548]: I first removed the `.map(FakeDefId::from)` part and needed to fix it later, which I used `Into::into`, but I agree that `FakeDefId::from` is better.

File: src/librustdoc/passes/collect_trait_impls.rs (hunk 3 above)
Comment (Stupremee) [reply src/librustdoc/passes/collect_trait_impls.rs:215]: it passed a reference before so I kept it, but passing by value probably makes more sense yeah

**Comment by jyn514** [comment]
LGTM but I don't have time to look closely at the code.

r? @GuillaumeGomez

File: src/librustdoc/html/render/write_shared.rs [hunk 4]
```diff
@@ -464,8 +464,6 @@ pub(super) fn write_shared(
     // Update the list of all implementors for traits
     let dst = cx.dst.join("implementors");
     for (&did, imps) in &cx.cache.implementors {
-        let did = did.expect_real();
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/write_shared.rs:467]: Why removing this?

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/write_shared.rs:467]: Ah nevermind, we changed `implementors` into `DefId`, so there is no need for that.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks!

@bors: r=jyn514,GuillaumeGomez

**Comment by Stupremee** [comment]
@GuillaumeGomez this error seems completely unrelated to my changes or am I missing something?

**Comment by GuillaumeGomez** [comment]
Indeed, let's retry!

@bors: retry

**Comment by GuillaumeGomez** [comment]
I think it'll need to be rebased on master:

```
    Checking tracing-subscriber v0.2.16
    Checking tracing-tree v0.1.9
    Checking rustdoc-json-types v0.1.0 (/checkout/src/rustdoc-json-types)
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0599]: no method named `as_local` found for enum `types::FakeDefId` in the current scope
  --> src/librustdoc/passes/check_code_block_syntax.rs:56:42
   |
56 |         let local_id = match item.def_id.as_local() {
   |                                          ^^^^^^^^ help: there is an associated function with a similar name: `is_local`
  ::: src/librustdoc/clean/types.rs:54:1
   |
   |
54 | crate enum FakeDefId {
   | -------------------- method `as_local` not found for this
error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustdoc`
```

**Comment by GuillaumeGomez** [comment]
Sorry, I made a mistake. It was another PR. But thanks for the rebase!

@bors: r=jyn514,GuillaumeGomez
